### PR TITLE
Implement cycleways

### LINKED ISSUE #139: Quest: cycleway
How about a quest for cycleways:
http://wiki.openstreetmap.org/wiki/DE:Key:cycleway

Are you able to use the images used in the above wiki article?

### PR DESCRIPTION
**Pull request by westnordost**
Fixes #139

For public review.

### CODE REVIEW

**Comment by matkoniecz** [comment]
Very impressive interface!

Nothing fits "road oneway for vehicles, not oneway for bicycles, no separate lane" normally tagged cycleway=opposite.

Nothing fits for "road oneway for vehicles, not oneway for bicycles, separate bicycle contraflow lane" - cycle lane option is displaying both general purpose lane (that may not exist in one direction) and cyclist lane.

maybe skip also surface=gravel and similar roads? http://www.openstreetmap.org/way/25116628#map=19/50.07320/19.88894 and similar are extremely unlikely to have cycleway infrastructure.

On oneway roads some answers make no sense and some may imply oneway:bicycle=no.

Open quest, rotate phone - question marks on unanswered parts are disappering (may affect also other parts of interface).

Would it be a good idea to have option to quickly select "no, there is no cycling infrastructure of any kind here"? Currently it requires tapping one direction, selecting none, tapping second direction, selecting none, selecting OK.

Generally, I am not sure whatever it makes sense to spam cycleway:both=no everywhere - but maybe it is result of 

- living in a city that is unfriendly to cyclists (so there is low amount of cycleways for tagging)
- living in areas with well mapped cycling infrastructure (so what exists is mapped)
- preferring mapping cycleways as separate ways

**Comment by matkoniecz** [comment]
Maybe skip also roads with low max speed? I think that maxspeed=20/30 roads are unlikely to have separate cycleways.

**Comment by westnordost** [comment]
> Nothing fits "road oneway for vehicles, not oneway for bicycles, no separate lane" normally tagged cycleway=opposite.

Hmm because this is not really what this quest asks for. It asks for cycleways, not whether a oneway-road is not oneway for bicycles.

**Comment by westnordost** [comment]
> Nothing fits for "road oneway for vehicles, not oneway for bicycles, separate bicycle contraflow lane" - cycle lane option is displaying both general purpose lane (that may not exist in one direction) and cyclist lane.

In that case, you are supposed to select simply what you see. I.e. you stand at the entrance of a oneway road and you see a cycleway lane on the left side (contraflow), you simply select "lane" for the left side (seen from your point of view) and "nothing" for right side. The quest automatically detects that this is indeed a contraflow-lane and tags it appropriately (`oneway:bicycle=no, cycleway:left=lane, cycleway:left:oneway=-1`).
See [AddCyclewayTest.java](https://github.com/westnordost/StreetComplete/pull/671/files#diff-abdee353badd7d3a3a37b1f9aead9b6b)

**Comment by westnordost** [comment]
> Open quest, rotate phone - question marks on unanswered parts are disappering (may affect also other parts of interface).

**BUG**

**Comment by westnordost** [comment]
> Would it be a good idea to have option to quickly select "no, there is no cycling infrastructure of any kind here"? Currently it requires tapping one direction, selecting none, tapping second direction, selecting none, selecting OK.

Hmm, I think it is fast enough as it is. It is basically 4 clicks vs. 2 clicks in that case.
I initially had a click-area in the middle of that street, when clicked, it would set both the left and the right side to the selected cycleway. But it was not self-explanatory enough and also too small to well distinguish from the other click areas.

**Comment by matkoniecz** [comment]
> you simply select "lane" for the left side (seen from your point of view) and "nothing" for right side

It was not obvious for me that 1) it is OK to leave empty part of answer 2) app will be smart enough to recognize it as contraflow.

**Comment by westnordost** [comment]
No, it is not OK to leave a part empty. "Nothing" is an option in the dialog. It says "No cycleway" or "none" or something in English.

> app will be smart enough to recognize it as contraflow.

You have no faith in me ;.-(

**Comment by matkoniecz** [comment]
>>  Nothing fits "road oneway for vehicles, not oneway for bicycles, no separate lane" normally tagged cycleway=opposite.

> Hmm because this is not really what this quest asks for. It asks for cycleways, not whether a oneway-road is not oneway for bicycles.

So app should add cycleway:both=none? Note that correct answer is cycleway:right=none, cycleway:left=opposite (or cycleway=opposite).

**Comment by matkoniecz** [comment]
> No, it is not OK to leave a part empty. "Nothing" is an option in the dialog. It says "No cycleway" or "none" or something in English.

So to tag "lane in one direction for everybody, contraflow lane in opposite direction" (cycleway=opposite_lane) one should select "cycle lane" option and "none" option?

This results in interface displaying road with bicycle lane in one direction and general-purpose lanes in both directions. App will add correct information but user is unaware about that.

There should be way to achieve situation where app displays only cycleway lane in one direction and only general-purpose lane in other direction.

**Comment by westnordost** [comment]
The app does not use the `opposite`-tagging, there is an alternative way to tag it, namely `oneway:bicycle=no`.
I created a forum discussion about that recently. Sadly, it's in the German forum. Perhaps you want to use Google Translate if you are interested: https://forum.openstreetmap.org/viewtopic.php?id=59810

Only OSM experts would know that when thinking in tags, "a oneway road without cycleways but where cyclists can go in both directions" would be some kind of "cycleway".

**Comment by westnordost** [comment]
> So to tag "lane in one direction for everybody, contraflow lane in opposite direction" (cycleway=opposite_lane) one should select "cycle lane" option and "none" option?

You mean there are cycleway lanes on both sides? Then, select both lane for left and for right. Just select what you see, do not worry about contraflow.

**Comment by matkoniecz** [comment]
>>  So to tag "lane in one direction for everybody, contraflow lane in opposite direction" (cycleway=opposite_lane) one should select "cycle lane" option and "none" option?

> You mean there are cycleway lanes on both sides? Then, select both lane for left and for right. Just select what you see, do not worry about contraflow.

No, there is only contraflow lane. Typical in Poland.

800px-kontrapas_poznan_ul_wroniecka

**Comment by matkoniecz** [comment]
> Only OSM experts would know that when thinking in tags, "a oneway road without cycleways but where cyclists can go in both directions" would be some kind of "cycleway".

Also for roads with explicit markings? Like in this case (especially as this symbol already appears as one of possibilities):

images duckduckgo com

**Comment by westnordost** [comment]
> first picture

right: cycleway lane
left: no cycleway

> second picture

right: explicit shared lane
left: no cycleway

**Comment by matkoniecz** [comment]
>>    first picture

> right: cycleway lane
> left: no cycleway

This results in interface displaying road with bicycle lane in one direction and general-purpose lanes in both directions. App will add correct information but user is unaware about that.

>>   second picture

>right: explicit shared lane
>left: no cycleway

"explicit shared lane" is named "lane shared explicitly with other traffic" despite that only bicycles are allowed in that direction.

Potential solutions:

add option "only cycle lane" (synonymous during edit to "cycle lane" but displayed differently in app, to avoid confusing users by showing general-purpose lanes in both directions for cases like first picture).

Rename "lane shared explicitly with other traffic" to "bicycles explicitly allowed without separate lane" to cover also cases like second picture.

**Comment by matkoniecz** [comment]
> The app does not use the opposite-tagging, there is an alternative way to tag it, namely oneway:bicycle=no.

> I created a forum discussion about that recently. Sadly, it's in the German forum. Perhaps you want to use Google Translate if you are interested: https://forum.openstreetmap.org/viewtopic.php?id=59810

AFAIK oneway:bicycle=no should be tagged in addition to cycleway/cycleway:both/cycleway:left/cycleway:right tags (and maybe it should be tagged for some answers in this quest).

**Comment by westnordost** [comment]
Ah now I get it. The picture always shows a normal road next to the selected cycleway. You are suggesting that this can lead to misunderstandings.

I guess a solution would be to have different pictures for when it is a oneway road perhaps? On the other hand, if the road has many more than 2 lanes, the pictures also do not change.

**Comment by westnordost** [comment]
Or, remove the center marking from the picture, so it is undefined how many lanes there are on the road, it is just a dark-gray surface.

**Comment by matkoniecz** [comment]
> I guess a solution would be to have different pictures for when it is a oneway road perhaps?

That may be a better solution than ideas from https://github.com/westnordost/StreetComplete/pull/671#issuecomment-341567330

Probably the best way to check is it necessary is to verify whatever others also would be confused (for example "if the road has many more than 2 lanes, the pictures also do not change" is not confusing for me).

> Or, remove the center marking from the picture, so it is undefined how many lanes there are on the road, it is just a dark-gray surface.

O! I think that it would work, at least for me.

File: app/src/main/java/de/westnordost/streetcomplete/quests/bikeway/AddCycleway.java [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+		   element became fit for this quest. */
+		return false;
+	}
+
+	@Override public boolean download(BoundingBox bbox, MapDataWithGeometryHandler handler)
+	{
+		return overpassServer.getAndHandleQuota(getOverpassQuery(bbox), handler);
+	}
+
+	/** @return overpass query string to get streets without cycleway info not near paths roads that don't have names */
```
Comment (matkoniecz) [inline app/src/main/java/de/westnordost/streetcomplete/quests/bikeway/AddCycleway.java]: > overpass query string to get streets without cycleway info not near paths roads that don't have names

is quite confusing, especially "roads that don't have names" given that there is no filtering by name that I see.

**Comment by westnordost** [comment]
@matkoniecz would you look again and see if it is better now in your opinion?

**Comment by matkoniecz** [comment]
> It is basically 4 clicks vs. 2 clicks in that case.

For me it was still irritating enough that I made a separate quest with higher priority that asks whatever there is any cycling infrastructure.

At least in my city in overwhelming majority of cases I answer cycleway=no.

**Pull request merged**